app/src/main/java/com/example/weather/MainActivity.java

```diff
@@ -1,14 +1,186 @@
+                    notificationOnError(context, getText(R.string.error_wrong_request).toString(), e);
+                    return;
+                }
+
+                String temperatureTemplate = new String((String) getText(R.string.template_temperature_message));
+                String temperatureMessage = String.format(temperatureTemplate, temperature);
+                updateTemperatureView(temperatureMessage);
+
+                Log.d(TAG, "run finish");
+            }
+        }).start();
+
+        Log.d(TAG, "onClick finish");
     }
+
+    public void onClickByLocation(View view) {
+        // todo: implement
+        // https://trello.com/c/W4VxNHog
+    }
+
+    private void updateTemperatureView(String massage) {
+        showWeatherView.post(new Runnable() {
+            @Override
+            public void run() {
+                showWeatherView.setText(massage);
+            }
+        });
+    }
+
+    private URL buildRequestUrlWithCity(String cityName) throws MalformedURLException {
+        Uri builtUri = Uri.parse(getText(R.string.weather_api_entry_point).toString())
+                .buildUpon()
+                .appendQueryParameter("q", cityName)
+                .appendQueryParameter("appid", getText(R.string.weather_api_key).toString())
+                .appendQueryParameter("units", "metric")
+                .build();
+        return new URL(builtUri.toString());
+    }
+
+    @Nullable
+    private String parseTemperature(String response) throws JSONException {
+        String temperature = null;
+
+        JSONObject json = new JSONObject(response);
+        JSONObject main  = json.getJSONObject("main");
+        temperature = main.getString("temp");
+
+        return temperature;
+    }
+
+    @Nullable
+    private String doRequest(URL weatherEndpoint) throws IOException {
+        Log.d(TAG, "doRequest start");
+        String respond = null;
+
+        InputStream stream = null;
+        HttpsURLConnection connection = null;
+
+        try {
+            connection = (HttpsURLConnection) weatherEndpoint.openConnection();
+            connection.setRequestMethod("GET");
+            connection.setReadTimeout(3000);
+            connection.connect();
+
+            if (connection.getResponseCode() == HttpsURLConnection.HTTP_OK) {
+                stream = connection.getInputStream();
+                respond = StreamUtils.StreamToString(stream);
+            }
+        } finally {
+            StreamUtils.closeAll(stream);
+            if (connection != null) {
+                connection.disconnect();
+            }
+        }
+        Log.d(TAG, "doRequest finish");
+        return respond;
+    }
+
+    private void notificationOnError(Context context, String notificationToUser) {
+        updateTemperatureView(getText(R.string.default_temperature_message).toString());
+        runOnUiThread(() -> Toast.makeText(context, notificationToUser, Toast.LENGTH_SHORT).show());
```

Comment:
runOnUiThread where is the definition of this method?

Reply:
runOnUiThread is the Activity's class method
https://developer.android.com/reference/android/app/Activity#runOnUiThread(java.lang.Runnable)

This expression makes possible to call Toast messenger from non mainUI thread